### PR TITLE
[SPARK-31935][SQL][3.0] Hadoop file system config should be effective in data source options

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Mkae Hadoop file system config effective in data source options.

From `org.apache.hadoop.fs.FileSystem.java`:
```
  public static FileSystem get(URI uri, Configuration conf) throws IOException {
    String scheme = uri.getScheme();
    String authority = uri.getAuthority();

    if (scheme == null && authority == null) {     // use default FS
      return get(conf);
    }

    if (scheme != null && authority == null) {     // no authority
      URI defaultUri = getDefaultUri(conf);
      if (scheme.equals(defaultUri.getScheme())    // if scheme matches default
          && defaultUri.getAuthority() != null) {  // & default has authority
        return get(defaultUri, conf);              // return default
      }
    }
    
    String disableCacheName = String.format("fs.%s.impl.disable.cache", scheme);
    if (conf.getBoolean(disableCacheName, false)) {
      return createFileSystem(uri, conf);
    }

    return CACHE.get(uri, conf);
  }
```
Before changes, the file system configurations in data source options are not propagated in `DataSource.scala`.
After changes, we can specify authority and URI schema related configurations for scanning file systems.

This problem only exists in data source V1. In V2, we already use `sparkSession.sessionState.newHadoopConfWithOptions(options)` in `FileTable`.
### Why are the changes needed?

Allow users to specify authority and URI schema related Hadoop configurations for file source reading.

### Does this PR introduce _any_ user-facing change?

Yes, the file system related Hadoop configuration in data source option will be effective on reading.

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
This PR backports https://github.com/apache/spark/pull/28760 to branch-3.0

**Comment by gengliangwang** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]
LGTM if tests pass

**Comment by cloud-fan** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Please hold on this PR because this will break Hadoop 3.2.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
@gengliangwang . Please include my follow-PR here.
- c7d45c0e0b8c077da8ed4a902503a6102becf255

**Comment by gengliangwang** [comment]
@dongjoon-hyun sure, I have included it in this PR. Thanks.

**Comment by dongjoon-hyun** [comment]
Thanks~

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 1]
```diff
@@ -532,6 +532,18 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
...
 
+  test("SPARK-31935: Hadoop file system config should be effective in data source options") {
+    withTempDir { dir =>
+      val path = dir.getCanonicalPath
+      val defaultFs = "nonexistFS://nonexistFS"
+      val expectMessage = "No FileSystem for scheme: nonexistFS"
+      val message = intercept[java.io.IOException] {
+        spark.readStream.option("fs.defaultFS", defaultFs).text(path)
+      }.getMessage
+      assert(message == expectMessage)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:543]: Oops. It seems that I missed here.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:543]: Could you fix this place and forward-port to `master`, too?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:543]: ok, let me do it now

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:543]: Thanks again!

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
I reverted https://github.com/apache/spark/commit/dad163f0f4a70b84f560de4f1f0bb89354e39217 from branch-3.0.

**Comment by cloud-fan** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you, @gengliangwang and all.
Merged to branch-3.0.

**Pull request closed without merging**